File: prisma/refs/java/minimum-window-substring.java
import java.util.*;

class Solution {
    public String minWindow(String s, String t) {
        if (t.length() > s.length()) return "";
        Map<Character, Integer> need = new HashMap<>();
        for (char c : t.toCharArray()) need.merge(c, 1, Integer::sum);
        Map<Character, Integer> window = new HashMap<>();
        int have = 0, required = need.size();
        int bestL = -1, bestR = -1, l = 0;
        for (int r = 0; r < s.length(); r++) {
            char c = s.charAt(r);
            window.merge(c, 1, Integer::sum);
            if (need.containsKey(c) && window.get(c).equals(need.get(c))) have++;
            while (have == required) {
                if (bestL == -1 || r - l < bestR - bestL) { bestL = l; bestR = r; }
                char lc = s.charAt(l);
                window.merge(lc, -1, Integer::sum);
                if (need.containsKey(lc) && window.get(lc) < need.get(lc)) have--;
                l++;
            }
        }
        return bestL == -1 ? "" : s.substring(bestL, bestR + 1);
    }
}

File: prisma/refs/py/minimum-window-substring.py
from collections import Counter


def min_window(s, t):
    if len(t) > len(s):
        return ""
    need = Counter(t)
    have, required = 0, len(need)
    window = {}
    best = None
    l = 0
    for r, c in enumerate(s):
        window[c] = window.get(c, 0) + 1
        if c in need and window[c] == need[c]:
            have += 1
        while have == required:
            if best is None or r - l < best[1] - best[0]:
                best = (l, r)
            lc = s[l]
            window[lc] -= 1
            if lc in need and window[lc] < need[lc]:
                have -= 1
            l += 1
    return s[best[0]:best[1] + 1] if best else ""
